fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/BaseMetadataAvailability.java

```diff
@@ -0,0 +1,56 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.table.availability;
+
+import com.yahoo.bard.webservice.data.config.names.DataSourceName;
+import com.yahoo.bard.webservice.metadata.DataSourceMetadataService;
+import com.yahoo.bard.webservice.table.resolver.PhysicalDataSourceConstraint;
+import com.yahoo.bard.webservice.util.SimplifiedIntervalList;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * An availability based on a DataSourceMetadataService, which is backed by a single data source in that service.
+ */
+public abstract class BaseMetadataAvailability implements Availability {
+
+    protected final DataSourceName dataSourceName;
+    protected final Set<DataSourceName> dataSourceNames;
+    protected final DataSourceMetadataService metadataService;
+
+    /**
+     * Constructor.
+     *
+     * @param dataSourceName  The name of the data source associated with this Availability
+     * @param metadataService  A service containing the datasource segment data
+     */
+    public BaseMetadataAvailability(
+            @NotNull DataSourceName dataSourceName,
+            @NotNull DataSourceMetadataService metadataService
+    ) {
+        this.metadataService = metadataService;
+        this.dataSourceName = dataSourceName;
+        this.dataSourceNames = Collections.singleton(dataSourceName);
+    }
+
+    public DataSourceName getDataSourceName() {
+        return dataSourceName;
+    }
+
+    @Override
+    public Set<DataSourceName> getDataSourceNames() {
+        return dataSourceNames;
```

Comment:
I guess I don't really care either way. I think caching the prewrapped collection doesn't introduce any real two-sources-of-truth issue because the thing being wrapped is final.

So I'm gently preferring the performance optimization of skipping the extra call to Collections.singlton (not on space or GC, but on a fast-loop performance basis)